File: src/main/java/discretemath/bool/expression/forms/SumOfProductExpression.java
package discretemath.bool.expression.forms;

import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import discretemath.bool.expression.BooleanExpression;
import discretemath.bool.expression.atomic.BooleanVariable;
import discretemath.bool.expression.atomic.BooleanVariables;
import discretemath.bool.expression.exception.VariableValueNotSpecifiedException;
import discretemath.bool.minimization.McCluskey;
import discretemath.bool.operator.FunctionallyCompleteOperatorsSet;
import discretemath.bool.operator.Operators;
import discretemath.bool.table.FunctionValue;
import discretemath.bool.table.TruthTable;
import discretemath.bool.table.TruthTableRow;

public class SumOfProductExpression implements BooleanExpression {

	private final List<BooleanVariable> variables;

	private final LinkedHashSet<Minterm> minterms;

	private SumOfProductExpression(List<BooleanVariable> variables, LinkedHashSet<Minterm> minterms) {
		this.variables = variables;
		this.minterms = minterms;
	}

	public static SumOfProductExpression fromTruthTable(TruthTable truthTable) {
		int functionsCount = truthTable.getFunctionsCount();
		if (functionsCount != 1) {
			throw new IllegalArgumentException("Truth table must be contain exactly one function");
		}

		int variablesCount = truthTable.getVariablesCount();

		List<BooleanVariable> variables = BooleanVariables.getVariables(variablesCount);

		LinkedHashSet<Minterm> minterms = new LinkedHashSet<>();
		for (TruthTableRow truthTableRow : truthTable) {
			if (truthTableRow.getFunctionValue(0) == FunctionValue.TRUE) {

				LinkedHashSet<Literal> literals = new LinkedHashSet<>();
				for (int i = 0; i < truthTableRow.argumentsCount(); i++) {
					boolean value = truthTableRow.getArgument(i);
					BooleanVariable variable = variables.get(i);
					literals.add(new SimpleLiteral(variable, value));
				}

				minterms.add(new SimpleMinterm(literals));
			}
		}

		return new SumOfProductExpression(variables, minterms);
	}

	public List<BooleanVariable> getVariables() {
		return variables;
	}

	public LinkedHashSet<Minterm> getMinterms() {
		return minterms;
	}

	@Override
	public boolean evaluate(Map<BooleanVariable, Boolean> arguments) throws VariableValueNotSpecifiedException {
		boolean[] values = new boolean[minterms.size()];

		Iterator<Minterm> mintermIterator = minterms.iterator();

		for (int i = 0; i < minterms.size(); i++) {
			Minterm minterm = mintermIterator.next();

			values[i] = minterm.evaluate(arguments);
		}

		return Operators.OR.evaluate(values);
	}

	@Override
	public BooleanExpression expressViaOperators(FunctionallyCompleteOperatorsSet operatorsSet) {
		throw new UnsupportedOperationException(); //TODO
	}

	@Override
	public BooleanExpression normalize() {
		return new McCluskey().minimizeSOPExpression(this);
	}

	@Override
	public String toString() {
		return minterms.stream().map(Object::toString).collect(Collectors.joining(" + "));
	}
}

File: src/main/scala/discretemath/bool/expression/forms/SumOfProductExpression.scala
package discretemath.bool.expression.forms

import discretemath.bool.expression.BooleanExpression
import discretemath.bool.expression.atomic.BooleanVariable
import discretemath.bool.expression.exception.VariableValueNotSpecifiedException
import discretemath.bool.minimization.McCluskey
import discretemath.bool.operator.FunctionallyCompleteOperatorsSet
import discretemath.bool.operator.Operators.OR
import discretemath.bool.table.{FunctionValue, TruthTable}

import scala.collection.immutable.ListSet

class SumOfProductExpression private(val variables: IndexedSeq[BooleanVariable], val minterms: Set[Minterm]) extends BooleanExpression {

  override def getVariables: Set[BooleanVariable] = minterms.flatMap(_.getVariables)

  @throws[VariableValueNotSpecifiedException]
  def evaluate(arguments: Map[BooleanVariable, Boolean]): Boolean = {
    val functionValuesCount = 1 << variables.length

    return minterms.size match {
      case 0 => false
      case `functionValuesCount` => true
      case _ => OR.evaluate(minterms.view.map(_.evaluate(arguments)).toSeq)
    }
  }

  def expressViaOperators(operatorsSet: FunctionallyCompleteOperatorsSet): BooleanExpression = OR.injectUsing(operatorsSet, minterms.toSeq)

  override def minimize: BooleanExpression = McCluskey.minimizeSOPExpression(this)

  override def isSatisfiable: Boolean = {
    return minterms.size match {
      case 0 => false
      case _ => true
    }
  }

  override def toString: String = minterms.view.map(_.toString).mkString(" + ")
}

object SumOfProductExpression {

  def apply(truthTable: TruthTable): SumOfProductExpression = {
    val variables = BooleanVariable.getN(truthTable.getVariablesCount)
    return apply(truthTable, variables)
  }

  def apply(truthTable: TruthTable, variables: IndexedSeq[BooleanVariable]): SumOfProductExpression = {
    val functionsCount = truthTable.getFunctionsCount
    if (functionsCount != 1) {
      throw new IllegalArgumentException("Truth table must be contain exactly one function")
    }

    val variablesCount = truthTable.getVariablesCount
    assert(variables.length == variablesCount, s"Variable list size expected: $variablesCount but was ${variables.length}")

    val minterms: ListSet[Minterm] = truthTable.view
      .filter(row => row.getFunctionValue(0) == FunctionValue.TRUE)
      .map(row => {
        val literals: ListSet[Literal] = variables.view
          .zip(row.arguments)
          .map(tuple => Literal(tuple._1, tuple._2))
          .to(ListSet)

        Minterm(literals)
      }).to(ListSet)

    return new SumOfProductExpression(variables, minterms)
  }
}